### ISSUE TITLE
Spine v3.8 json format breaking changes

### ISSUE THREAD

**Issue by YuryKuvetski** [pixijs/pixi-spine#300]
As  I understand in Spine v3.8 json format was changed, skins now is an Array.
It breaks your logic in the readSkeletonData function,  it expects that root.skins is an object with skin names as keys and attachments as values.

**Comment by ivanpopelyshev** [comment]
well shit. OK, I can make another pass on changes in spine . There's a way to speed-up this: do you know if there is skin example in official spine-ts repo that i can use to test if its fixed?

**Comment by YuryKuvetski** [comment]
is that enough?
http://esotericsoftware.com/spine-json-format

**Comment by ivanpopelyshev** [comment]
No. I'll take skins example from https://github.com/EsotericSoftware/spine-runtimes/blob/3.7/spine-ts/webgl/demos/skins.js

**Comment by YuryKuvetski** [comment]
@ivanpopelyshev Also there is a bug with a bones section of an animation.
```
"bones": {
   "boneName": {
      "timelineType": [
         { "time": 0, "angle": -26.55 },
         { "time": 0.1333, "angle": -8.78 },
...
```
**"time"** parameter is optional now as I understand.
So in my case it looks like this:
```
"animations": {
  "animation": {
    "bones": {
      "_0008_male_main_body": {
        "rotate": [
          { "curve": 0.25, "c3": 0.75 },
          { "time": 0.8667, "angle": 1.92, "curve": 0.25, "c3": 0.75 },
          { "time": 1.6667 }
        ],
        "translate": [
          { "curve": 0.25, "c3": 0.75 },
          { "time": 0.8333, "x": -29.15, "y": -216.47, "curve": 0.25, "c3": 0.75 },
          { "time": 1.6667 }
        ],
        "scale": [
          { "curve": 0.25, "c3": 0.75 },
          { "time": 0.8333, "x": 1.133, "y": 1.133, "curve": 0.25, "c3": 0.75 },
          { "time": 1.6667 }
        ]
      },
...
```
So time is undefined and after some calculations it is NaN. Solution is to use 0 if time is undefined.

**Comment by ivanpopelyshev** [comment]
How do I update pixi-spine: I take all commits between previous and current version of https://github.com/EsotericSoftware/spine-runtimes/tree/3.8-beta/spine-ts , and apply them to pixijs runtime.

Thanks to your notices i'll be able to do it faster ;)

**Comment by anna8810** [comment]
Hi, @ivanpopelyshev will these changes be added to the 4.x branches as well?

**Comment by ivanpopelyshev** [comment]
Only by demand :) Is your project big enough to stay on v4 because of possible problems with migration?

**Comment by anna8810** [comment]
Yes, that would make us miss our deadline... :(

**Comment by ivanpopelyshev** [comment]
Spine 3.8 is still in beta. Are you sure you need it?

**Comment by anna8810** [comment]
Our animator unfortunately updated to the beta version, and did a lot of work that I need to get in.

**Comment by ivanpopelyshev** [comment]
OK, start the countdown

**Comment by djmisterjon** [comment]
@anna8810 you do the same error as me lol.
badlogic and nate tell me how get back to preview and no lost your projet.
http://esotericsoftware.com/forum/3-8-13-crash-pixijs-and-corrupted-project-file-Search-this-12066?p=53931#p53931

**Comment by ivanpopelyshev** [comment]
Well, i've got sidetracked by emergency, but im working on it. In a pair of hours you will get pixi-v4 + spine3.8

**Comment by ivanpopelyshev** [comment]
OK, please test this branch: https://github.com/pixijs/pixi-spine/tree/v4.x-3.8beta/bin

I've tested only raptor demo, but i didnt test any skins features

**Comment by djmisterjon** [comment]
work fine on my side , thanks for this

**Comment by ivanpopelyshev** [comment]
@YuryKuvetski @anna8810 guys?

**Comment by anna8810** [comment]
Sorry, we had a public holiday in Sweden yesterday so I was not at work. 
This will work for me, thanks guys! <3

**Comment by YuryKuvetski** [comment]
Works fine for me, thanks. Sorry for late.

**Comment by YuryKuvetski** [comment]
@ivanpopelyshev Should your code support old spine formats?
![image](https://user-images.githubusercontent.com/20362835/59276661-6d9c7100-8c67-11e9-8da1-be16cf0c2dca.png)

I can't find any logic which can parse an old format of skins. So I'm getting this exception(
![image](https://user-images.githubusercontent.com/20362835/59276939-e6033200-8c67-11e9-8eff-70fd062c4149.png)

**Comment by ivanpopelyshev** [comment]
No, its supposed to be 3.8-beta only, because there are too many changes that I dont know if its actually possible to load something old and show it the same way as new one.

**Comment by ivanpopelyshev** [comment]
OK, both v4 and v5 pixi + spine 3.7 3.8 are supported now. Just in different branches.

**Comment by Ashden** [comment]
> No, its supposed to be 3.8-beta only, because there are too many changes that I dont know if its actually possible to load something old and show it the same way as new one.

It is possible, I've cloned v4.x-3.8beta and updated the changes to the JSON export, and tested on a spine containing slots, skins, constraints, transforms and blend mode. It is backward-compatible with 3.7 and no regression issues for 3.8 (I can provide an example export from both 3.7 and 3.8 for demo purposes)

@ivanpopelyshev I read you said that is 3.8-beta only, so I'm wondering if you want a pull request into a specific branch, or a new branch for this?

Note: the changes are for PIXI v4 (which we are currently using), but I assume the changes are easily applicable to PIXI v5 (I can do that as well, if you want the backward-compatibility)

**Comment by ivanpopelyshev** [comment]
@Ashden Please look in the README - https://github.com/pixijs/pixi-spine#pixi-spine 

3.8beta branch is outdated.

3.8 has changes in IK and animations too, im not sure that they are not breaking.

When there were breaking changes between spine3.2 and 3.3 I also wondered whether i can make universal JSON import. The answer was "yes, but i dont want to support it in future".

Here is the same - there are already enough pixi-specific changes in runtime, and if I add some in JSON importer i'll take responsibility for it . I'm not sure that I'll be able to maintain them in future.

**Comment by Ashden** [comment]
@ivanpopelyshev If your plan is to only use the latest Spine implementation and use the same strategy as they do (which is the obvious and natural choice), then yes, I fully agree with your decision.
That being the case, I guess I can maintain a backwards-compatible version until all our projects get up to date with Spine 3.8, then we can move on your stable build for it.

PS: Just for kicks, I applied my patch on pixi4-spine3.8 and it's still working 😄

**Comment by ivanpopelyshev** [comment]
ok, give me the patch, just in case ;)

**Comment by ivanpopelyshev** [comment]
Of course I can make spine runtime much better than official, but.. ugh.. i have too many pixijs plugins on me :) need to stop sometimes. And badlogic might look at me funny, I dont want to undermine his authority on whats good for spine users.

Its better if I put more effort in publishing GPUPATCH again, it affects performance when people use big number of spine animations with deformed meshes.

**Comment by Ashden** [comment]
[backward-compatibility_to_3_7.patch.zip](https://github.com/pixijs/pixi-spine/files/3481601/backward-compatibility_to_3_7.patch.zip)

I have the issue with many spines with deformed meshes too, so I'd like a fix for that 😄

**Comment by ayanguptapcg** [comment]
Hi, I am trying to move to pixi 5 and spine 3.8.72. I have a very basic spine file and it is not working. Can someone please look into my files and see where I am wrong. This is very urgent as we are going to start a big game and we have already made a lot of animations using spine 3.8.72. Any help would be really appreciated.
[Spine_Run_tinkered.zip](https://github.com/pixijs/pixi-spine/files/3644635/Spine_Run_tinkered.zip)

**Comment by ivanpopelyshev** [comment]
@ayanguptapcg i dont understand what did you put in that whip.json , how it should work. Also `spineboy.json` is clearly outdated so if you tried to test that too - it wont work, its like very very very old json file without `skeleton` section.

**Comment by ayanguptapcg** [comment]
It gives AtlasAttachmentLoader.ts:43 Uncaught Error: Region not found in atlas: error

**Comment by ayanguptapcg** [comment]
For this error we made some change as demo file in whip atlas so atleast that was not throwing error but also does not show up anything.

**Comment by ivanpopelyshev** [comment]
Animated alpha is 0.03 . Any idea why?

**Comment by ivanpopelyshev** [comment]
nope, still nothing even if i set alpha to 1 everywhere

**Comment by ayanguptapcg** [comment]
are you checking the whip or mysteryDoll? I don't see animated alpha anywhere...

**Comment by ivanpopelyshev** [comment]
doll. i've modified atlas file of course.

**Comment by ayanguptapcg** [comment]
I am not able to see any alpha value in the atlas.. what should we do then? is 3.8.72 not supported? or are we doing something wrong. Can I get one 3.8 spine file which is working with the library to cross check or some hints for exports.

**Comment by ivanpopelyshev** [comment]
3.8 release with pixi-v5 works for everyone. Of course you are doing something different than other people. I dont know about that `.72` part, i didnt look at latest-latest changes. The problem is that I cant even determine whether that `wrong` is a spine bug , pixi-spine bug or just you messing with files :)

**Comment by ivanpopelyshev** [comment]
I'll try later again, but right now I suggest to look in official runtime  https://github.com/EsotericSoftware/spine-runtimes/tree/3.8/spine-ts and try to put your files in official spine widget, I remember that some people posted videos on youtube on how to do it.

If your stuff doesnt work in official widget too, then its a problem for esotericsoftware forum.

**Comment by ayanguptapcg** [comment]
Thanks Ivan.. we will do that and come back to you with our finding. Thanks a lot for taking this up.

**Comment by ivanpopelyshev** [comment]
Hold on, im picking up something strange on my side.

**Comment by ayanguptapcg** [comment]
ok.. fingers crossed

**Comment by ivanpopelyshev** [comment]
I published new version with a few changes that i picked from , but it doesnt help in your case. 

The bigger problem is: https://pixijs.io/examples/#/plugin-spine/dragon.js is for spine 3.7 , and I dont have ANY examples that run with pixi in 3.8 on this computer. I cant test anything. I dont even know whether current version works at all on demo files.

**Comment by ayanguptapcg** [comment]
I just ran a test with the runtime and it does not load there. I tried both. Whip loads but does not show any animation and doll does not load at all. will try with some more atlases.

**Comment by ivanpopelyshev** [comment]
Heh, then you have to bother guys from esotericsoftware :)

**Comment by ivanpopelyshev** [comment]
use `setSkin`, i guess :)

**Comment by ayanguptapcg** [comment]
Thank you Ivan..

**Comment by dhartiste** [comment]
Thanks for all this work!
I am setting up a new implementation of pixi-spine with pixi4 and Spine 3.8. It looks like that is an active project so wondering if it's possible for you to make an npm version of that. So I can do npm install pixi-spine@pixi4-spine3.8 or something similar.

**Comment by ivanpopelyshev** [comment]
I honestly dont know how to `npm publish` that without changing current latest version. I'm too lazy to view. Do you have a tutorial or something?

I updated pixi4-3.8 just today, with new build in `dist` folder.

**Comment by dhartiste** [comment]
In the meantime I found that you did have v1.6.1 which is a fairly recent build in npm for the pixi4 / spine3.8. You could publish the one you just updated with https://docs.npmjs.com/cli/v7/commands/npm-publish The syntax that they mention is below. It looks like as long as you use a tag then it would not be set to the 'latest'

`npm publish [<tarball>|<folder>] [--tag <tag>] [--access <public|restricted>] [--otp otpcode] [--dry-run]`

Publishes '.' if no argument supplied
Sets tag 'latest' if no --tag specified

**Comment by ivanpopelyshev** [comment]
Done, looks like it works:https://www.npmjs.com/package/pixi-spine?activeTab=versions

```bash
npm install pixi-spine@pixi4-spine3.8
```
